### ISSUE TITLE
spec2nii does not support raw data from United Imaging Healthcare(UIH)

### ISSUE THREAD

**Issue by longfangithub** [wexeee/spec2nii#1]
UIH is a MR manufacturer from China, and by now many institutes over the world have been using UIH MR scanners to acquire MRI data. However, the spec2nii currently does not support MRS raw data conversion obtained from UIH scanners. 
What information is required for MRS data format conversion? Maybe, I can provide some support to implement UIH data conversion. If and questions, please give me a reply or contact me via email fan.long@united-imaging.com

**Comment by wexeee** [comment]
Hi @longfangithub, sorry for the delay in replying, I've just got back from a couple of weeks of holiday. I've not yet come across any data from a UIH scanner, but I'm very happy to include it.

What sort of format does it normally come in? Does it contain voxel position information? Does it contain the absolute minimum metadata required (dwell time and central frequency)?

**Comment by longfangithub** [comment]
Thank you for the reply. The raw data format of United Imaging Healthcare is different from that of other manufacturers. MRS's raw data contains the information of voxel position information, dwell time and central frequency.
I think there are two ways for the conversion of data information. The first way is that we provide you with tools to parse the UIH's raw data. Second is that we provide tool that converts raw data into ismrmrd format. Which one do you think is more convenient?

**Comment by wexeee** [comment]
If there was a software tool for parsing and some example data that would be a great place to start. Are you interested in single voxel or CSI? Do you have some data to share which has multiple orientations from the scanner? What I've found is screenshots from the scanner console are very useful in debugging orientation issues.

**Comment by longfangithub** [comment]
We are interested in both single voxel and CSI, and I think we can start the work from single voxel. Next, we will prepare some test data and pass it to you.

**Comment by longfangithub** [comment]
Here we provide an example of two SVS data from UIH, including the T1w images and SVS data that was converted to ismrmrd format. And the script can be used to parse ismrmrd data.
The data is large, and I will email the data to william.clarke@ndcn.ox.ac.uk.

**Comment by longfangithub** [comment]
I have sent some sample data via email to william.clarke@ndcn.ox.ac.uk, did you receive it? Is there any problem with the data?

**Comment by wexeee** [comment]
Hi @longfangithub, Please accept my apologies, I took too long to download the data and it seems to have expired on the hosting system. Could you re-upload the files?

**Comment by longfangithub** [comment]
It doesn't matter. I've re-uploaded the files.

**Comment by longfangithub** [comment]
have you received the data?

**Comment by wexeee** [comment]
Yes I did , thanks. I haven’t had a chance to delve into this for a fix.

**Comment by longfangithub** [comment]
I have a suggestion that we implement the code and you review and verify it. Is this feasible?

**Comment by wexeee** [comment]
Hi @longfangithub spec2nii is currently undergoing a rewrite to make it compliant with the new NIfTI MRS standard. My attention is currently focused on that (and ISMRM). I would definitely like to include this format once that effort is done. However, it would be great if you could provide an e.g. jupyter notebook of how to read the complex data, the meta-data parameters from the file and also how to interpret any of the orientation information. This would substantially speed up inclusion.

**Comment by wexeee** [comment]
Hi @longfangithub Just to say I've made some progress on this.

For the data you've provided I can generate the following view in fsleyes.

![image](https://user-images.githubusercontent.com/22190542/102601946-bf556f80-4118-11eb-8e42-cee3df0d097f.png)

It would be great to have some test data that we are able to share widely. E.g. on a phantom and also some CSI if you have some data available.

Do you have any information on how to interpret the data in 'uprotocol' (DICOM tag 0065,1007) ? Or any of the private fields?

This is all in the nifti_mrs branch which should be completed in the next few days.

**Comment by longfangithub** [comment]
There are some explanation of common tag in DICOM on this page: 
https://github.com/rordenlab/dcm2niix/issues/225.
What other fields do you need? It is best to provide a list, and we will try our best to explain.

We will prepare some test data on phantom and some CSI data next week.

**Comment by wexeee** [comment]
I'll take a look through that thread on dcm2niix.

Nibabel reported the following spectroscopy specific private tags:
```
(0065, 00ff) Private Creator                     LO: 'Spectroscopy Private Header'
(0065, 1000) Private tag data                    UL: 1592475108
(0065, 1002) Private tag data                    LO: 'SAG > COR0.1'
(0065, 1003) Private tag data                    LO: 'HC1~3'
(0065, 1007) Private tag data                    OB: Array of 15188 elements
(0065, 100f) Private tag data                    SH: 'R28.1'
(0065, 1023) Private tag data                    LO: '0'
(0065, 1030) Private tag data                    LO: 'UIH All rights reserved'
(0065, ff01) Private tag data                    DS: [-38.0938034, -3.35821962, -4.38352728]
(0065, ff02) Private tag data                    DS: [0.085478887, 28.747551, 19.9642429, -0.00103385642, 5.70408869, -8.21360874, 19.9999409, -0.0389400013, -0.0295599997]
(0065, ff04) Private tag data                    DS: [35, 10, 20]
(0065, ff05) Private tag data                    DS: [35, 10, 20]
(0065, ff06) Private tag data                    DS: [-0.99999702, 0.00194700004, 0.00147799996, '']
(0065, ff07) Private tag data                    OF: Array of 16384 elements
(0065, ff08) Private tag data                    OF: Array of 16384 elements
(5600, 0020) Spectroscopy Data                   OF: Array of 16384 elements
```
(0065, ff08)  seems to contain very similar data to (5600, 0020). And (0065, ff07) contains water unsuppressed data.

ff01 to ff06 look like position and orientation information.

Is (0065, 1003) coil information? And is there a simple way of parsing (0065, 1007)? Perhaps in a similar way to the Siemens CSA header.

Phantom data would be great, especially if you have a phantom with some structure e.g. something like the GE resolution phantom. If you could get 2D and 3D CSI that would be ideal.

**Comment by longfangithub** [comment]
![mmexport1608708352421](https://user-images.githubusercontent.com/32872738/102970954-6da34f80-4533-11eb-9bf9-40dfea23b284.png)
The description of the private tag is as shown above. 
The way of parsing (0065, 1007) is different from that of CSA header. （0065, 1007) is the data describing the sequence, and its related content can be founded in the common tag of dicom.

**Comment by wexeee** [comment]
Hi @longfangithub,

I've been doing a substantial rewrite of the orientation calculation, mostly for Siemens but I'm using much of the same code for UIH data. Of the examples you've given me everything seems to match up nicely now except the 3D CSI where there is an inversion in the 3rd dimension. I notice that the orientation and PlaneOrientation parameters that you describe above also have a negation in them.
(0065, ff02) Private tag data                    DS: [160, -0, 0, 0, 160, 0, 0, 0, -100]
(0065, ff04) Private tag data                    DS: [160, 160, 100]
(0065, ff06) Private tag data                    DS: [0, 0, -1, '']

I imagine that these are therefore essential to parse to calculate the orientation information. Do you have any further information on (0065, ff06) specifically?

**Comment by wexeee** [comment]
Actually one more thing. The 2D CSI example you provided also seems to have some oddness in the 3rd dimension. The screenshot clearly shows the slab being bisected by the image at 0.0 mm H-F/S-I but the image position patient field in the corresponding DICOM file has a value of 7.5 (which by definition?) should relate to the slab centre. Is there some special treatment of this in the sequence code?

**Comment by longfangithub** [comment]
(0065, ff06) represents the normal vector of the layer direction. In UIH the main direction of the normal vector is defined to be negative. 

The image position(patient) refers to the coordinates of the first pixel in the upper left corner in the patient coordinate system, and it is the same as the DICOM standard.

**Comment by wexeee** [comment]
Hi @longfangithub I believe this is now fix and working as I would expect as of 3d463187974a2388458e886125536f0a36d01750. Therefore I'm closing this. Let me know if it works for you (I'll make a new release shortly).

**Comment by longfangithub** [comment]
It's great that the spec2nii works on UIH dicom data now. Maybe due to the new nifti standard, mrs_vis (fsl_mrs) can not work on the new nifti format?
And we also look forward that spec2nii can support for ismrmrd format in the future.
